parlai/core/worlds.py

```diff
@@ -1499,6 +1533,622 @@ def shutdown(self):
+        # QueueWorld init
+        self.world: World = world
+        self.agents: List[Agent] = world.get_agents()
+        self.processes: List[Process] = []
+        self.consume_queues: List[Queue] = []
+        self.worlds: List[World] = []
+        self.produce_queue: Queue = Queue()
+        self.report_queue: Queue = Queue()
+        self.buffers: List[Dict[str, torch.Tensor]] = []
+        self.finished_workers: List[int] = []
+        self.training = 'train' in opt['datatype'] and 'evalmode' not in opt['datatype']
+        self.numworkers = opt['numworkers'] if self.training else 1
+        self.init_parley = True
+        self.num_consume = 0
+        self._init_workers(opt, world, pworld_class)
+
+    def num_examples(self):
+        """
+        Return the number of examples for the root world.
+        """
+        return self.world.num_examples()
+
+    def num_episodes(self):
+        """
+        Return the number of episodes for the root world.
+        """
+        return self.world.num_episodes()
+
+    def get_total_exs(self):
+        """
+        Return the total number of processed episodes in the root world.
+        """
+        return self.world.get_total_exs()
+
+    def getID(self):
+        """
+        Return the ID of the root world.
+        """
+        return self.world.getID()
+
+    def get_agents(self):
+        """
+        Return the agents of the root world.
+        """
+        return self.world.get_agents()
+
+    def get_task_agent(self):
+        """
+        Return task agent of the root world.
+        """
+        return self.world.get_task_agent()
+
+    def get_model_agent(self):
+        """
+        Return model agent of the root world.
+        """
+        return self.world.get_model_agent()
+
+    def episode_done(self):
+        """
+        Return whether the episode is done.
+
+        A batch world is never finished, so this always returns `False`.
+        """
+        return False
+
+    def epoch_done(self):
+        """
+        Override to account for finished workers.
+        """
+        subworlds_done = self.world.epoch_done() or all(
+            w.epoch_done() for w in self.worlds
+        )
+        return subworlds_done or len(self.finished_workers) == self.numworkers
+
+    def save_agents(self):
+        """
+        Save the agents in the root world.
+        """
+        self.world.save_agents()
+
+    def _init_workers(self, opt: Opt, world: World, pworld_class: type):
+        """
+        Initialize the processes.
+
+        :param opt:
+            Opt dictionary
+
+        :param world:
+            Master world
+
+        :param pworld_class:
+            Which PWorld to use.
+        """
+        for worker_idx in range(self.numworkers):
+            consume_queue: Queue = Queue()
+            shared = world.share()
+            subworld = shared['world_class'](opt, None, shared)
+            self.processes.append(
+                Process(
```

Comment:
Yes so you can inherent from Process and implement the run function. It tends to read cleaner since you introduce class variables etc